### ISSUE TITLE
Contract interface is not refreshed upon redeployment

### ISSUE THREAD

**Issue by alxflw** [onflow/flow-playground#206]
**Describe the bug**
If a developers has a working contract and script but later on redeploys the contract, the script view doesn't seem to load the new interface. As a result, methods that might not work anymore don't show any syntax error.

**To Reproduce**
- Deploy a working contract
- Run a script against the contract
- Update the contract interface and redeploy
- open the script again (unaltered). The new interface isn't loaded. The code seems to look ok
- Run the script and it returns an error
- Reload the page and a syntax error will show up (assuming that the new contract interface is now loaded correctly)

**Expected behavior**
The contract interface is auto-reloaded upon redeployment and syntax errors show up

**Comment by MaxStalker** [comment]
I have investigated the issue and here are my findings:
- we create new language server for every template editor - `if the server for one editor crashes, it does not break the other editors` - not sure if that's affects the ability to share updated state between them :thinking: 
- I found there is a `getAddressCode` we set for language client callbacks. But calling it explicitly doesn't do the trick. @turbolent @sideninja  maybe you can help with this one?

**Comment by MaxStalker** [comment]
Feels like LanguageServer is caching response from `getAddressCode` :thinking: 

Adding a `console.log` inside `getAddressCode` callback and then switching import address from 0x01 -> 0x02 -> 0x03 -> 0x02 -> 0x01 -> 0x03 will only show 2 or 3 calls(on initial load) and not 5 or 6.

**Comment by MaxStalker** [comment]
Could be somewhere around here?
https://github.com/onflow/cadence/blob/1e61ffa65cd977f1baab73a36dd639ac6bcd4bf2/languageserver/cmd/languageserver/main_wasm.go#L161-L167

**Comment by MaxStalker** [comment]
@kerrywei can you grab someone with LanguageServer knowledge for help on this one?

**Comment by turbolent** [comment]
>  we create new language server for every template editor - if the server for one editor crashes, it does not break the other editors - not sure if that's affects the ability to share updated state between them

You are probably right that we probably missed this workflow when implementing the editors to each have their own language server, instead of having one which maintains multiple open documents, just like it is done in e.g. VS Code.

I don't remember the implementation details, but if it is indeed the case that each editor/document is backed by one language server, it is likely that there is no syncing across them, leading to the behaviour described above. In that case it would probably be best to refactor the Playground to use one language server instance across multiple editors, or add the missing synchronization code.

**Comment by MaxStalker** [comment]
As for `missing synchronization code`, @turbolent  - would it be something on language server side or I can send some message from client? :thinking:

**Comment by turbolent** [comment]
I think that the synchronization code would be basically involve listening for  changes in each editor, then sending [`textDocument/didChange`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_didChange) events to all language servers, so that they get the updated code. 

I haven't looked at any of this in a long time, so this might be totally wrong.

**Comment by MaxStalker** [comment]
@turbolent @kerrywei I found the problem with non-updated code was actually related to delayed `project` prop updates. `getAddressCode` would use closure around `project` variable and try to pull old data out of it, even though it was already updated by contract deployment.

But now we have a new problem. Let's say we have this code on account 0x01:
```kotlin
pub contract Basic{}
```
And a script, which imports it:
```kotlin
import Basic from 0x01

pub fun main(){}
```

If we redeploy the contract to 0x01:
```kotlin
pub contract Simple{}
```
And switch back to script - it will not highlight `Basic` as non-existing contract. When we start typing - language server `will` notify us that contract is non available on that account, unless we write `Simple`. At that point of type wavy red line and errors will disappear. The problem is - if we put it back to `Basic` - there will be no highlight either :fearful:

Feels like either client or server is caching version of the file and not resubmitting diagnostics for files it already checked.
On one hand - that's a smart feature, but we need a way to override it somehow. Maybe by submitting new file `version` after redeploy :shrug:

**Comment by MaxStalker** [comment]
Could be this capability? :thinking: 
https://microsoft.github.io/language-server-protocol/specification#textDocument_synchronization_sc

```
/**
 * Defines how the host (editor) should sync document changes to the language
 * server.
 */
export namespace TextDocumentSyncKind {
	/**
	 * Documents should not be synced at all.
	 */
	export const None = 0;

	/**
	 * Documents are synced by always sending the full content
	 * of the document.
	 */
	export const Full = 1;

	/**
	 * Documents are synced by sending the full content on open.
	 * After that only incremental updates to the document are
	 * send.
	 */
	export const Incremental = 2;
}```

**Comment by kerrywei** [comment]
> @turbolent @kerrywei I found the problem with non-updated code was actually related to delayed `project` prop updates. `getAddressCode` would use closure around `project` variable and try to pull old data out of it, even though it was already updated by contract deployment.
> 
> But now we have a new problem. Let's say we have this code on account 0x01:
> 
> ```kotlin
> pub contract Basic{}
> ```
> 
> And a script, which imports it:
> 
> ```kotlin
> import Basic from 0x01
> 
> pub fun main(){}
> ```
> 
> If we redeploy the contract to 0x01:
> 
> ```kotlin
> pub contract Simple{}
> ```
> 
> And switch back to script - it will not highlight `Basic` as non-existing contract. When we start typing - language server `will` notify us that contract is non available on that account, unless we write `Simple`. At that point of type wavy red line and errors will disappear. The problem is - if we put it back to `Basic` - there will be no highlight either 😨
> 
> Feels like either client or server is caching version of the file and not resubmitting diagnostics for files it already checked. On one hand - that's a smart feature, but we need a way to override it somehow. Maybe by submitting new file `version` after redeploy 🤷

thank you for the update @MaxStalker . would you want to further discuss this with some team members via a meeting? freel free to schedule one if more thoughts/ideas are needed

**Comment by MaxStalker** [comment]
I was thinking to rebuild npm package for language server locally and try it - just to make sure it's not the cause.

We can also opt for a hacky solution, which would reload the page, when contract is deployed. It would fix the problem, but it's not the most elegant one 🤷‍♂️

**Comment by MaxStalker** [comment]
As I was typing this I thought we could restart language server (and probably client) when contract is deployed... 🤔

**Comment by MaxStalker** [comment]
I found working combination of restarting language clients and server, when new contract was deployed.
The issue should be fixed when #222 will be merged into `main` branch :)

**Comment by kerrywei** [comment]
@MaxStalker could you provide status update on this task?

**Comment by kerrywei** [comment]
@MaxStalker is the change verified?

**Comment by MaxStalker** [comment]
@kerrywei it's deployed to production, but I see similar strange bugs. So can't say it's 100% fixed :(

**Comment by MaxStalker** [comment]
We have identified the problem and refactored the code to prevent this error from happening again.
During the process we've also been able to spot similar problem in VS Code Extension and then pin point location, where caching is happening on Language Server. This helped us improve performance and remove need to restart language server every time user would switch editor. New version of `@onflow/cadence-language-server` is not yet released, but by the when PR will be merged it should be on npm :crossed_fingers: 

Current work is gathered under this PR - #238
